### PR TITLE
Initial support for the Keyboardio Atreus

### PR DESCRIPTION
**Pull request by algernon**
This is a quick patch to add support for the upcoming [Keyboardio Atreus](https://www.kickstarter.com/projects/keyboardio/atreus). This is a straightforward patch, except there's no `info.json`, because I don't have the capacity at the moment to craft a layout (hence why this PR is a draft). I may - and likely will - tweak it a bit in the next few days.

I did want to get a draft PR going, so folks who are interested in the Keyboardio Atreus, but want to use QMK with it, will have a starting point.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/atreus_kbio/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+#include QMK_KEYBOARD_H
+
+// Each layer gets a name for readability, which is then used in the keymap matrix below.
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+// Layer names don't all need to be of the same length, obviously, and you can also skip them
+// entirely and just use numbers.
+#define _QW 0
+#define _RS 1
+#define _LW 2
```
Comment (noroadsleft) [inline keyboards/atreus_kbio/keymaps/default/keymap.c]:
```suggestion
enum layer_names {
    _QW,
    _RS,
    _LW,
};
```

File: keyboards/atreus_kbio/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+#define ___ KC_TRNS
+
```
Comment (noroadsleft) [inline keyboards/atreus_kbio/keymaps/default/keymap.c]:
```suggestion
```

File: keyboards/atreus_kbio/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  /*
+   *  !       @     up     {    }        ||     pgup    7     8     9    *
+   *  #     left   down  right  $        ||     pgdn    4     5     6    +
+   *  [       ]      (     )    &        ||       `     1     2     3    \
+   * lower  insert super shift bksp ctrl || alt space   fn    .     0    =
+   */
+  [_RS] = LAYOUT( /* [> RAISE <] */
+    KC_EXLM, KC_AT,   KC_UP,   KC_LCBR, KC_RCBR,                   KC_PGUP, KC_7,    KC_8,   KC_9, KC_ASTR ,
+    KC_HASH, KC_LEFT, KC_DOWN, KC_RGHT, KC_DLR,                    KC_PGDN, KC_4,    KC_5,   KC_6, KC_PLUS ,
+    KC_LBRC, KC_RBRC, KC_LPRN, KC_RPRN, KC_AMPR, ___,     ___,     KC_GRV,  KC_1,    KC_2,   KC_3, KC_BSLS ,
```
Comment (noroadsleft) [inline keyboards/atreus_kbio/keymaps/default/keymap.c]:
```suggestion
    KC_LBRC, KC_RBRC, KC_LPRN, KC_RPRN, KC_AMPR, _______, _______, KC_GRV,  KC_1,    KC_2,   KC_3, KC_BSLS ,
```

File: keyboards/atreus_kbio/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  /*
+   * insert home   up  end   pgup       ||      up     F7    F8    F9   F10
+   *  del   left  down right pgdn       ||     down    F4    F5    F6   F11
+   *       volup             reset      ||             F1    F2    F3   F12
+   *       voldn  super shift bksp ctrl || alt space   L0  prtsc scroll pause
+   */
+  [_LW] = LAYOUT( /* [> LOWER <] */
+    KC_INS,  KC_HOME, KC_UP,   KC_END,  KC_PGUP,                   KC_UP,   KC_F7,   KC_F8,   KC_F9,   KC_F10  ,
+    KC_DEL,  KC_LEFT, KC_DOWN, KC_RGHT, KC_PGDN,                   KC_DOWN, KC_F4,   KC_F5,   KC_F6,   KC_F11  ,
+    KC_NO,   KC_VOLU, KC_NO,   KC_NO,   RESET,   ___,     ___,     KC_NO,   KC_F1,   KC_F2,   KC_F3,   KC_F12  ,
```
Comment (noroadsleft) [inline keyboards/atreus_kbio/keymaps/default/keymap.c]:
```suggestion
    KC_NO,   KC_VOLU, KC_NO,   KC_NO,   RESET,   _______, _______, KC_NO,   KC_F1,   KC_F2,   KC_F3,   KC_F12  ,
```

File: keyboards/atreus_kbio/readme.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Keyboardio Atreus
+
+![Atreus](https://atreus.technomancy.us/i/atreus2-mug.jpg)
+
+A small mechanical keyboard that is based around the shape of the human hand.
+
+Keyboard Maintainer: [Keyboardio](https://github.com/keyboardio)
+Hardware Supported: Keyboardio Atreus
+Hardware Availability: https://www.kickstarter.com/projects/keyboardio/atreus
```
Comment (noroadsleft) [inline keyboards/atreus_kbio/readme.md]:
```suggestion
* Keyboard Maintainer: [Keyboardio](https://github.com/keyboardio)
* Hardware Supported: Keyboardio Atreus
* Hardware Availability: https://www.kickstarter.com/projects/keyboardio/atreus
```

File: keyboards/atreus_kbio/rules.mk [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# MCU name
+MCU = atmega32u4
+
+# Bootloader selection
+BOOTLOADER = caterina
```
Comment (noroadsleft) [inline keyboards/atreus_kbio/rules.mk]:
```suggestion
#   Teensy       halfkay
#   Pro Micro    caterina
#   Atmel DFU    atmel-dfu
#   LUFA DFU     lufa-dfu
#   QMK DFU      qmk-dfu
#   ATmega32A    bootloadHID
#   ATmega328P   USBasp
BOOTLOADER = caterina
```

**Comment by algernon** [comment]
> I think I'd prefer this go to `keyboardio/atreus` instead of `atreus_kbio`, but that's simply my opinion.

I was considering that, but there's no `keyboardio/` subdir yet. I'm ok with moving things to `keyboardio/atreus` though. If that's the recommendation, I'll do that, and then followup later with moving the `model01` to `keyboardio/model01` too.

**Comment by algernon** [comment]
I went ahead and force-pushed the branch, with @noroadsleft's suggestions included, and the keyboard moved to `keyboards/keyboardio/atreus`.

Come to think of it, it might make sense to do the `model01`->`keyboardio/model01` move _first_. I'll prep a PR for that, and will rebase this on top of that.

**Comment by algernon** [comment]
...and now it's rebased on top of #8499.

**Comment by drashna** [comment]
Personally, I'd rather it not be rebased on the other PR, as that will definitely slow down/block this from being merged.  

I do agree with using the /keyboardio/ folder though.

**Comment by algernon** [comment]
Well, there's no `info.json` yet, and I do not have the resources to create one in the near future, so this is going to be blocked for a while anyway, no?

I can rebase on master and just use the keyboardio dir, though, that makes sense. I'll push that in a bit.

**Comment by noroadsleft** [comment]
> Well, there's no `info.json` yet, and I do not have the resources to create one in the near future, so this is going to be blocked for a while anyway, no?

I can probably do this by the end of today and send a PR to your fork.

**Comment by noroadsleft** [comment]
How's this?

image

https://github.com/noroadsleft/qmk_firmware/commit/4f461693378215bf43bc70915ab027e18cd301be

**Comment by algernon** [comment]
That looks very nice. Thank you!

**Comment by drashna** [comment]
Yeah, noroads is our info.json master. :D

File: keyboards/keyboardio/atreus/rules.mk [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
+BOOTLOADER = caterina
```
Comment (noroadsleft) [inline keyboards/keyboardio/atreus/rules.mk]: This line makes me sad. :cry:

**Comment by drashna** [comment]
@algernon okay with this being merged, as is?

**Comment by technomancy** [comment]
Thanks for putting this together! I've tested it on my board and everything works great, except for the reset functionality. The reset key causes it to disconnect and reconnect, but it does not enter the bootloader.

File: keyboards/keyboardio/atreus/rules.mk (hunk 7 above)
Comment (fauxpark) [reply keyboards/keyboardio/atreus/rules.mk:12]: Most likely this board just has atmel-dfu on it, which would make sense with @technomancy's comment below.

Comment (noroadsleft) [inline keyboards/keyboardio/atreus/rules.mk:12]:
```suggestion
BOOTLOADER = atmel-dfu
```

Comment (drashna) [reply keyboards/keyboardio/atreus/rules.mk:12]: > All ATmega32U4 chips save for two come with atmel-dfu pre-installed. Sparkfun reflashes them with Caterina (😠) for use on Pro Micros.

And PJRC flashes them with Halfkay, and locks the bootloader. 😀

**Comment by technomancy** [comment]
Oh hang on; I realized this is actually based on the old Atreus Classic layout rather than the new Keyboardio Atreus one; the parens should be on either side of the fn arrow cluster, but they are below the arrow cluster instead.

**Comment by technomancy** [comment]
I can't figure out how to do a PR to @algernon 's branch, so I'll just drop this commit here which updates it to use the fn layer that matches what this board ships with: https://github.com/technomancy/qmk_firmware/commit/3284d7776e3f5e300e760ad3425bc593bb2193a2

The reset issue is still a problem, however given the accessible reset button on the reverse it's not as severe of a problem as it would be on the classic Atreus.

**Comment by algernon** [comment]
Sorry for the inactivity, I'll merge both the info.json and @technomancy's changes too ~tomorrow.

As for the bootloader: the Keyboardio Atreus ships with Catelina, a custom version of that too. To enter bootloader mode, one needs to hold the bottom left key (`Esc` by default), even if reset via a keypress. So holding `Esc` and then pressing the reset macro (or the other way around, if you're fast enough) should put you into bootloader mode.

**Comment by noroadsleft** [comment]
> To enter bootloader mode, one needs to hold the bottom left key (`Esc` by default), even if reset via a keypress. So holding `Esc` and then pressing the reset macro (or the other way around, if you're fast enough) should put you into bootloader mode.

This should absolutely be stated in the main readme.

> the Keyboardio Atreus ships with [Caterina], a custom version of that too.

This line makes me very sad. :cry:

**Comment by algernon** [comment]
Picked @noroadsleft's `info.json`, and @technomancy's layout update, and updated the `readme.md`.

File: keyboards/keyboardio/atreus/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+   *  !       @     up     {    }        ||     pgup    7     8     9    *
+   *  #     left   down  right  $        ||     pgdn    4     5     6    +
+   *  [       ]      (     )    &        ||       `     1     2     3    \
+   * lower  insert super shift bksp ctrl || alt space   fn    .     0    =
+   */
+  [_RS] = LAYOUT( /* [> RAISE <] */
+    KC_EXLM, KC_AT,   KC_UP,   KC_DLR,  KC_PERC,                  KC_PGUP, KC_7,    KC_8,   KC_9, KC_BSPC,
+    KC_LPRN, KC_LEFT, KC_DOWN, KC_RGHT, KC_RPRN,                  KC_PGDN, KC_4,    KC_5,   KC_6, KC_BSLS,
+    KC_LBRC, KC_RBRC, KC_HASH, KC_LPRN, KC_RPRN, KC_CIRC, KC_AMPR,KC_ASTR, KC_1,    KC_2,   KC_3, KC_PLUS,
+    TG(_L2), KC_INS,  KC_LGUI, KC_LSFT, KC_BSPC, KC_LCTL, KC_LALT,KC_SPC,  KC_TRNS, KC_DOT, KC_0, KC_EQL
```
Comment (noroadsleft) [inline keyboards/keyboardio/atreus/keymaps/default/keymap.c:32]: Required change; keymap doesn't build without it.

- `_L2` doesn't exist
- missing closing parenthesis, so layer doesn't terminate

```suggestion
    TG(_LW), KC_INS,  KC_LGUI, KC_LSFT, KC_BSPC, KC_LCTL, KC_LALT,KC_SPC,  KC_TRNS, KC_DOT, KC_0, KC_EQL
  ),
```

File: keyboards/keyboardio/atreus/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  /*
+   *  !       @     up     {    }        ||     pgup    7     8     9    *
+   *  #     left   down  right  $        ||     pgdn    4     5     6    +
+   *  [       ]      (     )    &        ||       `     1     2     3    \
+   * lower  insert super shift bksp ctrl || alt space   fn    .     0    =
+   */
+  [_RS] = LAYOUT( /* [> RAISE <] */
+    KC_EXLM, KC_AT,   KC_UP,   KC_DLR,  KC_PERC,                  KC_PGUP, KC_7,    KC_8,   KC_9, KC_BSPC,
+    KC_LPRN, KC_LEFT, KC_DOWN, KC_RGHT, KC_RPRN,                  KC_PGDN, KC_4,    KC_5,   KC_6, KC_BSLS,
+    KC_LBRC, KC_RBRC, KC_HASH, KC_LPRN, KC_RPRN, KC_CIRC, KC_AMPR,KC_ASTR, KC_1,    KC_2,   KC_3, KC_PLUS,
```
Comment (noroadsleft) [inline keyboards/keyboardio/atreus/keymaps/default/keymap.c:31]: @technomancy Seems to me the parentheses on this row should actually be curly brackets, based on [this image from the Kickstarter page](https://ksr-ugc.imgix.net/assets/028/440/943/a0cc2331c9bf601b387ce86eebfb1fcd_original.png?ixlib=rb-2.1.0&w=680&fit=max&v=1584312206&auto=format&frame=1&lossless=true&s=87a9b1e3fc55034fd5dffa98c881165e).

```suggestion
    KC_LBRC, KC_RBRC, KC_HASH, KC_LCBR, KC_RCBR, KC_CIRC, KC_AMPR,KC_ASTR, KC_1,    KC_2,   KC_3, KC_PLUS,
```

Also the block comment on Lines 23-26 is no longer accurate.

**Comment by algernon** [comment]
...that's what I get for not testing locally. Fix incoming when I wake up.

**Comment by drashna** [comment]
> This line makes me very sad. 😢

I agree.  It would have been nice if QMK DFU was included on it, since a custom bootloader was used. 



> ...that's what I get for not testing locally. Fix incoming when I wake up.

lol, been bitten by that, more than once. :D

**Comment by bleleet** [comment]
@algernon 
Hi~ I want this issue to be active again.
I made atreus firmware without knowing the existence of this issue. 
I'm kickstarter early backer. I got a keyboard yesterday. After testing, QMK firmware works well.

**Comment by drashna** [comment]
@algernon any update on this?

**Comment by shieldsd** [comment]
Is this likely to be re-opened, if not shall I create a new pull request?

**Comment by evantravers** [comment]
I'm willing to bet we should revisit this… since the keyboardio shipments are going to start arriving.

**Comment by noroadsleft** [comment]
I have a copy of this branch saved locally; It would be trivial for me to rebase and update it, but I would very much like someone to test it as I don't/won't have the hardware.

**Comment by evantravers** [comment]
I rebased algernon:keyboardio/atreus on qmk:master locally and I ran into errors with `make keyboardio/atreus:default`.

I'll have the hardware this Monday… as long as I'm not in grave danger of bricking it I'll give it a whirl for science. 👍

**Comment by noroadsleft** [comment]
@evantravers Applying the changes in https://github.com/qmk/qmk_firmware/pull/8492#discussion_r406860161 should make it compile.

**Comment by evantravers** [comment]
@noroadsleft… ok… that compiles. I'll have it over the weekend, and then I need to some reading about what I would have to do to get QMK's firmware on that bootloader… I may have to swap bootloaders even? I'm a little out of my league there, but I'll google around and give it a good go.

**Comment by noroadsleft** [comment]
> I need to some reading about what I would have to do to get QMK's firmware on that bootloader… I may have to swap bootloaders even?

Thus saith `keyboards/keyboardio/atreus/readme.md` (emphasis added):

> To flash new firmware, the keyboard needs to be in bootloader mode. To do that,
one needs to **hold the bottom left key (`Esc` by default) and reset the
keyboard** - either via a macro, or by sending a hang-up signal to the keyboard
device, or **by pressing the reset button on the bottom with a pin**.

:slightly_smiling_face:

**Comment by evantravers** [comment]
Typing on a Keyboardio Atreus powered by QMK. 

image

running `make keyboardio/atreus:default:flash`, then pressing the lower ESC key and hitting the reset button worked just fine.

What are the next steps towards getting this in the main branch?

File: keyboards/keyboardio/atreus/config.h [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+
+#define VENDOR_ID       0x1209
```
Comment (Erovia) [inline keyboards/keyboardio/atreus/config.h:23]: I'm a bit concerned about this VID.
It's already used by a few boards with VIA support:
https://github.com/the-via/keyboards/blob/master/src/duckle/vitamins_included.json#L3
https://github.com/the-via/keyboards/blob/master/src/jiran/jiran.json#L3

(The fact that 2 VIA-enabled board already uses it is another topic.)

File: keyboards/keyboardio/atreus/config.h [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+#define PRODUCT_ID      0x2303
+#define DEVICE_VER      0x0000
+#define MANUFACTURER    Keyboardio
+#define PRODUCT         Atreus
+#define DESCRIPTION     QMK keyboard firmware for the Keyboardio Atreus
```
Comment (Erovia) [inline keyboards/keyboardio/atreus/config.h:28]:
```suggestion
```
As per https://docs.qmk.fm/#/pr_checklist

File: keyboards/keyboardio/atreus/readme.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+* Hardware Availability: https://www.kickstarter.com/projects/keyboardio/atreus
+
+These configuration files are specifically for the Keyboardio Atreus, a
+collaboration between [Phil Hagelberg](https://github.com/technomancy) and
+[Keyboardio](https://github.com/keyboardio). The keyboard is currently on
+Kickstarter, and will come fully assembled.
+
+Make example for this keyboard (after setting up your build environment):
+
+    make keyboardio/atreus:default:avrdude
```
Comment (Erovia) [inline keyboards/keyboardio/atreus/readme.md:18]:
```suggestion
    make keyboardio/atreus:default
    
  Flashing example for this keyboard:
  
    make keyboardio/atreus:default:flash
```

File: keyboards/keyboardio/atreus/rules.mk [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
```
Comment (Erovia) [inline keyboards/keyboardio/atreus/rules.mk:17]:
```suggestion
BOOTMAGIC_ENABLE = lite       # Virtual DIP switch configuration
```

**Review by Erovia — CHANGES_REQUESTED** [review]

**Review by Erovia — CHANGES_REQUESTED** [review]
Without noroadsleft's suggestions, does not build.

**Comment by evantravers** [comment]
@Erovia: I have a copy of the branch with @noroadsleft's suggestions... is there an acceptable way for me to make your changes and swap out the branch so that this can get done, or is that impolite to @algernon who did the initial work?

File: keyboards/keyboardio/atreus/config.h (hunk 10 above)
Comment (tzarc) [reply keyboards/keyboardio/atreus/config.h:23]: 1209 is the VID for pid.codes -- it's properly registered: http://pid.codes/1209/2303/

**Comment by evantravers** [comment]
I created a PR for my adjusted branch… if that's rude or not kosher I am very sorry… just trying to get this in master to help all those new atreus owners. :P

**Comment by jmorag** [comment]
What version of avrdude does this use, @evantravers? I'm on 6.3. Tried running `make keyboardio/atreus:default:flash` on your branch and am getting
```
avrdude: Error: butterfly programmer uses avr_write_page() but does not
provide a cmd() method.
 *** page 125 (addresses 0x5d7e - 0x5dfd) failed to write

Writing | ################################################## | 100% 9.74s

avrdude: 24062 bytes of flash written
avrdude: verifying flash memory against .build/keyboardio_atreus_default.hex:
avrdude: load data flash data from input file .build/keyboardio_atreus_default.hex:
avrdude: input file .build/keyboardio_atreus_default.hex auto detected as Intel Hex
avrdude: input file .build/keyboardio_atreus_default.hex contains 24062 bytes
avrdude: reading on-chip flash data:

Reading | #                                                  | 2% 0.00savrdude: butterfly_recv(): programmer is not responding
Reading | ##                                                 | 4% 5.01savrdude: butterfly_recv(): programmer is not responding
Reading | ################################################## | 100% 10.16s

avrdude: verifying ...
avrdude: verification error, first mismatch at byte 0x0000
         0xff != 0x0c
avrdude: verification error; content mismatch

avrdude: safemode: Fuses OK (E:CB, H:D8, L:FF)

avrdude done.  Thank you.

make[1]: *** [tmk_core/avr.mk:318: flash] Error 1
Make finished with errors
make: *** [Makefile:579: keyboardio/atreus:default:flash] Error 1
```

**Comment by evantravers** [comment]
@jmorag I'm using `avrdude version 6.3` as well. Make sure you have checked out the right branch… I know mine works, but the changes requested haven't been made to the branch on this PR.

**Comment by Xyverz** [comment]
I can confirm that Evan's branch works for me. I'll go add my comment to his other PR, if it's still open.

**Comment by algernon** [comment]
Closing, #10570 is the successor of this, and that's been merged, so this one's really obsolete.

**Comment by rrix** [comment]
> Tried running `make keyboardio/atreus:default:flash` on your branch and am getting

(sorry to comment on a closed PR, but this got me, too so i'm gonna leave a breadcrumb)

try to disable ModemManager on your system or apply the udev rules referenced in #8060 . it's mentioned in the upstream Keyboardio/Chrysalis docs and maybe should be added to the device QMK readme.

**Comment by skullydazed** [comment]
This is also a situation `qmk doctor` will warn about. You can use that to check if you're affected.

**Pull request closed without merging**